src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala

```diff
@@ -676,6 +681,35 @@ trait SyncControllerBuilder {
 
 }
 
+trait PortForwardingBuilder {
+  self: DiscoveryConfigBuilder =>
+
+  import monix.execution.Scheduler.Implicits.global
+
+  private val portForwardingRelease = new AtomicReference(Option.empty[Future[Task[Unit]]])
+
+  def startPortForwarding(): Unit =
+    portForwardingRelease.compareAndSet(
+      None,
+      Some {
+        PortForwarder
+          .openPorts(
+            Seq(Config.Network.Server.port),
+            Seq(discoveryConfig.port).filter(_ => discoveryConfig.discoveryEnabled)
+          )
+          .whenA(Config.Network.automaticPortForwarding)
+          .allocated
+          .runToFuture
+          .map(_._2)
+      }
+    )
```

Comment:
I think this isn't doing the same thing as my version with the `Promise` was: `compareAndSet` takes both arguments by value, so the `Some` part will be evaluated even if the value is already set to something, and as a side effect start another port forwarder if the method is called twice in a row. Whereas with the `Promise` based version, if there already was a value, then all that happened was that we discarded a never-used `Promise` instance. 

```scala
@ import java.util.concurrent.atomic.AtomicReference 
import java.util.concurrent.atomic.AtomicReference

@ import scala.concurrent.ExecutionContext.Implicits.global 
import scala.concurrent.ExecutionContext.Implicits.global

@ import scala.concurrent.Future 
import scala.concurrent.Future

@ val ref: AtomicReference[Option[Future[Unit]]] = new AtomicReference(None) 
ref: AtomicReference[Option[Future[Unit]]] = None

@ ref.compareAndSet(None, Some(Future{ println("Hello!") })) 
res4: Boolean = true

Hello!
@ ref.compareAndSet(None, Some(Future{ println("Hello!") })) 
Hello!
res5: Boolean = false
```

See the 2nd invocation of `compareAndSet` returned `false` but still the side effect was executed. What you want is _only_ carry out the action if the return value was `true`.

Reply:
Ah yeah you're absolutely right. Going to improve this.

The atomic reference might as well contain a boolean btw, with the promise being a separate value. (I dislike how side-effecty all this stuff is...)

Reply:
Yeah it's unfortunate. I was wondering if it should be wrapped up in a helper; so far it didn't need reuse, but it's tricky to get right. 

The reason I thought the `Promise` to be in the reference was so that it can be replaced with a fresh one upon restart. Also because I could always tell that the system isn't running by the absence of it (which you don't need here).